common/workflow-core/src/main/scala/org/apache/texera/service/util/S3StorageClient.scala

```diff
@@ -139,4 +141,118 @@ object S3StorageClient {
       s3Client.deleteObjects(deleteObjectsRequest)
     }
   }
+
+  /**
+    * Uploads an object to S3 using multipart upload.
+    * Handles streams of any size without loading into memory.
+    */
+  def uploadObject(bucketName: String, objectKey: String, inputStream: InputStream): String = {
+    val buffer = new Array[Byte](MINIMUM_NUM_OF_MULTIPART_S3_PART.toInt)
+
+    // Helper to read a full buffer from input stream
+    def readChunk(): Int = {
+      var offset = 0
+      var read = 0
+      while (
+        offset < buffer.length && {
+          read = inputStream.read(buffer, offset, buffer.length - offset); read > 0
+        }
+      ) {
+        offset += read
+      }
+      offset
+    }
+
+    // Read first chunk to check if stream is empty
+    val firstChunkSize = readChunk()
+    if (firstChunkSize == 0) {
+      return s3Client
+        .putObject(
+          PutObjectRequest.builder().bucket(bucketName).key(objectKey).build(),
+          RequestBody.fromBytes(Array.empty[Byte])
+        )
+        .eTag()
+    }
+
+    val uploadId = s3Client
+      .createMultipartUpload(
+        CreateMultipartUploadRequest.builder().bucket(bucketName).key(objectKey).build()
+      )
+      .uploadId()
+
+    try {
+      // Upload all parts using an iterator
+      val allParts = Iterator
+        .iterate((1, firstChunkSize)) { case (partNum, _) => (partNum + 1, readChunk()) }
+        .takeWhile { case (_, size) => size > 0 }
+        .map {
+          case (partNumber, chunkSize) =>
+            val eTag = s3Client
+              .uploadPart(
+                UploadPartRequest
+                  .builder()
+                  .bucket(bucketName)
+                  .key(objectKey)
+                  .uploadId(uploadId)
+                  .partNumber(partNumber)
+                  .build(),
+                RequestBody.fromBytes(buffer.take(chunkSize))
+              )
+              .eTag()
+            CompletedPart.builder().partNumber(partNumber).eTag(eTag).build()
+        }
+        .toList
+
+      s3Client
+        .completeMultipartUpload(
+          CompleteMultipartUploadRequest
+            .builder()
+            .bucket(bucketName)
+            .key(objectKey)
+            .uploadId(uploadId)
+            .multipartUpload(CompletedMultipartUpload.builder().parts(allParts.asJava).build())
+            .build()
+        )
+        .eTag()
+
+    } catch {
+      case e: Exception =>
+        try {
+          s3Client.abortMultipartUpload(
+            AbortMultipartUploadRequest
+              .builder()
+              .bucket(bucketName)
+              .key(objectKey)
+              .uploadId(uploadId)
+              .build()
+          )
+        } catch { case _: Exception => }
```

Comment:
instead of having the nested try. catch for abort, will a `final {}` at the outer block make the code better? In the final you should be able to handle any exception come out of the above code blocks in try-catch